### DISCUSSION TITLE
isatty and capsys/capfd

### DISCUSSION THREAD

**Discussion by danizen** [pytest-dev/pytest#9525] [Q&A]
I've written a class to enhance the progress module providing my own "ChunkedSpinner" and "DotCounter".  To properly test these, I should make sure they are printing respectively a spinning character or a dot every 50 calls to report progress - this is the point, to make sure that progress reporting doesn't make everything slower.

However, the progress module uses the isatty call to prevent output, and I'm wondering how that will play with capsys.  For now, I am just testing that my code doesn't blow up and that the count of progression remains the same even with the chunking.  Any suggestions?

**Comment by portyu9** [comment]
Hi @danizen 

This is expected: pytest's captured `stdout`/`stderr` streams are **not real terminal devices**, so `isatty()` returns false.

`capsys` and `capfd` capture writes; they do not emulate a terminal.

For code whose behavior intentionally changes when connected to a TTY, I would make that dependency injectable and test it with a small fake:

```python
import io


class FakeTTY(io.StringIO):
    def isatty(self):
        return True


def test_progress():
    stream = FakeTTY()

    show_progress(stream=stream)

    assert "." in stream.getvalue()
```

Alternatively, monkeypatch the particular `isatty()` check your application uses.

If you genuinely need to test OS-level terminal behavior rather than your application's decision logic, use a pseudo-terminal (`pty`) on platforms that support it.

Please mark this answer as accepted if it helped, thank you.
